plugins/poll/lib/irv.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+class DiscoursePoll::Irv
+  def self.irv_outcome(poll_id)
+    options = PollOption.where(poll_id: poll_id).map { |hash| { id: hash.digest, html: hash.html } }
+
+    ballot = []
+
+    #Fetch all votes for the poll in a single query
+    votes =
+      PollVote
+        .where(poll_id: poll_id)
+        .select(:user_id, :poll_option_id, :rank)
+        .order(:user_id, :rank)
+        .includes(:poll_option) # Eager load poll options
+    # Group votes by user_id
+    votes_by_user = votes.group_by(&:user_id)
+    # Build the ballot
+    votes_by_user.each do |user_id, user_votes|
+      ballot_paper =
+        user_votes.select { |vote| vote.rank > 0 }.map { |vote| vote.poll_option.digest }
+      ballot << ballot_paper
+    end
+
+    DiscoursePoll::Irv.run_irv(ballot, options) if ballot.length > 0
+  end
+
+  def self.run_irv(starting_votes, options)
+    current_votes = starting_votes.dup
+    round_activity = []
+    round = 0
+    while true
```

Comment:
Can't we have a bounded loop? (if we really need this, the idiomatic ruby for this is `loop do`)

Reply:
you can, good idea! https://github.com/discourse/discourse/pull/27155/commits/09e19615012768f66210e5ca2c56a47a3fb74ce7